src/client/datascience/jupyter/kernels/kernelExecution.ts

```diff
@@ -91,117 +95,189 @@ export class KernelExecution implements IDisposable {
-            this,
-            this.disposables
-        );
+            .forEach((cell) => this.createCellExecution(cell));
 
         try {
-            const notebook = await notebookPromise;
-            const kernel = this.getKernel(document);
-            codeCellsToExecute.forEach((exec) => traceCellMessage(exec.cell, 'Ready to execute'));
-            for (const cellToExecute of codeCellsToExecute) {
-                traceCellMessage(cellToExecute.cell, 'Before Execute individual cell');
-                const result = this.executeIndividualCell(kernel, cellToExecute, notebook);
-                result.finally(() => this.cellExecutions.delete(cellToExecute.cell)).catch(noop);
-                const executionResult = await result;
-                traceCellMessage(cellToExecute.cell, `After Execute individual cell ${executionResult}`);
-                // If a cell has failed or execution cancelled, the get out.
-                if (
-                    cancelTokenSource.token.isCancellationRequested ||
-                    executionResult === vscodeNotebookEnums.NotebookCellRunState.Error
-                ) {
-                    traceInfo(
-                        `Cancel all remaining cells ${cancelTokenSource.token.isCancellationRequested} || ${executionResult}`
-                    );
-                    await Promise.all(
-                        codeCellsToExecute.map((cell) => {
-                            // Cancel and remove from map pending cells.
-                            return this.cancelCell(cell.cell);
-                        })
-                    );
-                    break;
-                }
-            }
+            await this.pendingExecution.chainExecution(async () => this.executeQueuedCells(notebookPromise, document));
         } finally {
-            traceInfo(`Cancel all remaining cells after finally`);
-            await Promise.all(
-                codeCellsToExecute.map((cell) => {
-                    // Cancel and remove from map pending cells.
-                    return this.cancelCell(cell.cell);
-                })
-            );
             this.documentExecutions.delete(document);
             traceInfo('Restore notebook state to idle');
             await editor.edit((edit) =>
                 edit.replaceMetadata({ ...document.metadata, runState: vscodeNotebookEnums.NotebookRunState.Idle })
             );
         }
     }
-
-    public async cancelCell(cell: NotebookCell) {
-        const execution = this.cellExecutions.get(cell);
-        if (execution) {
-            this.cellExecutions.delete(cell);
-            traceCellMessage(cell, 'Cancel cell from Kernel Execution');
-            await execution.cancel();
-        } else {
-            traceCellMessage(cell, 'Cannot cancel cell execution from Kernel Execution');
+    public async interrupt(notebookPromise: Promise<INotebook>, document: NotebookDocument): Promise<InterruptResult> {
```

Comment:
Moved from CellExecution, else we ended up with multiple interrupts & things got into a bad state.